feat: Add common matrix functions to standard library

keenancrane opened the pull request:
<img width="1254" alt="image" src="https://github.com/penrose/penrose/assets/3604525/1fccea31-4110-404b-b00c-03e6439c4614">

This PR adds common matrix functions, including 2D/3D transformations and camera matrices, to the library.  It also adds corresponding documentation, tests and examples, and a language feature `then` for left-to-right composition of transformations.

### New functions

New functions added to `Functions.ts` are listed below.  See the new `Vectors and Matrices` page in the Documentation, as well as the generated Function Library documentation, for further details.  (Design decisions for the naming conventions are also spelled out int his documentation.)

* `fromHomogeneous`
* `fromHomogeneousList`
* `toHomogeneous`
* `toHomogeneousList`
* `toHomogeneousMatrix`
* `rotate`
* `scale`
* `skew`
* `shear`
* `translate`
* `rotate2d`
* `scale2d`
* `skew2d`
* `shear2d`
* `rotate3d`
* `scale3d`
* `shear3d`
* `rotate3dh`
* `scale3dh`
* `shear`
* `translate3dh`
* `lookAt`
* `perspective`
* `ortho`
* `project`
* `projectDepth`
* `projectList`
* `matrixMultiplyList`

### New language features

This PR adds the `then` keyword to the Style parser/compiler.  This keyword is equivalent to the `*` operator, except that it reverses the order of the arguments, i.e., `A then B` is equilvalent to `B * A`.  The main use case, as highlighted in the documentation, is to compose transformations in "natural language order", i.e., left to right.  For instance, just as one can say "flip it over and put it in the oven" to mean "flip first, put in oven second," one can also write

```haskell
translate(x,y) then rotate(θ) then scale(a,b)
```

### Tests and examples

The `triangle-mesh-3d` example has been updated to use the new 3D transformations and camera matrices.

A new example, `dinoshade`, has also been added, with an extensive README that serves as a "Style for OpenGL Programmers" guide.

All the new matrix functions come with a suite of corresponding tests, in the `examples/matrix-library/`. These examples use the "red/green" testing paradigm, as explained in `examples/matrix-library/README.md`.  The particular tests are

* `crossProductMatrix.trio.json`
* `diagonal2d.trio.json`
* `diagonal3d.trio.json`
* `identity2d.trio.json`
* `identity3d.trio.json`
* `inverse2d.trio.json`
* `inverse3d.trio.json`
* `matrix2d.trio.json`
* `matrix3d.trio.json`
* `outerProduct2d.trio.json`
* `outerProduct3d.trio.json`
* `rotate.trio.json`
* `rotate2d.trio.json`
* `rotate3d.trio.json`
* `rotate3dh.trio.json`
* `scale2d.trio.json`
* `scale3d.trio.json`
* `shear2d.trio.json`
* `shear3d.trio.json`
* `skew2d.trio.json`
* `translate2d.trio.json`
* `translate3dh.trio.json`

### Documentation

As noted above, extensive documentation for new functionality has been added in

1. the inline function descriptions (hence the automatically generated documentation of library functions),
2. the new `Vectors and Matrices` page of the documentation, which has also refactored documentation on vector/matrix types and operations from the Style Usage page, and
3. the `dinoshade` example, which provides a step-by-step tutorial on usage of 3d transformation and camera matrices.

wodeni reviewed (changes requested):
Thanks for adding these functions and the great examples @keenancrane! Really appreciated the effort to document and test them. I'm sure they will enable many more diagrams soon!

I pushed one commit to tweak the documentation, and have some questions about `vouter` and error-handling. Hoping you can explain the former and @liangyiliang can possibly confirm the latter.

liangyiliang commented on packages/core/src/contrib/Functions.ts:
```diff
@@ -6035,3 +6956,779 @@ const diffusionProcess = (
+      mul(A[0][3], C30),
+    );
+  } else {
+    throw Error("matrix must be 2x2, 3x3, or 4x4");
```
If a Style function throws an error, it does _not_ give a return value. In other words the program flow is terminated entirely. Warnings get propagated to the top, but do not terminate the program. In this case throwing errors is what I would do

keenancrane commented:
@wodeni I believe I've fixed your requested changes, and also updated the vim highlighting files with my local ones (the ones in the repo were long out of date…).  Let me know if there's anything else!

wodeni reviewed (approved):
Look great. thanks @keenancrane!
